src/store/data_layer/data_store.js

```diff
@@ -0,0 +1,207 @@
+      .filter((key) => parseInt(key) >= payload.start && parseInt(key) <= payload.end) // Filter out irrelevant times
+      .sort((a, b) => parseInt(a) > parseInt(b)) // Chronologically sort times
+      .filter((key, index, array) => {
+        // Keep the first, the last, and any indices defining the start and/or end of a gap
+        return index === 0 || index === array.length - 1 || Math.abs(key - array[index - 1]) > 900 || Math.abs(key - array[index + 1]) > 900
+      })
+      .map(ts => parseInt(ts)) // Parse them to integers
+
+    // If no matching records are found, return the original interval
+    if (timestamps.length === 0) return [[payload.start, payload.end]]
+    else {
+      // At least one timestamp was found
+      // At this point, we have an array of all intervals where we have data stored
+      // We have the start and end times for each interval (inclusive)
+      // Now, we need to determine if there are any gaps.
+      let returnArr = [] // An array of interval pairs, ie: [[start, end]]
+      if (timestamps[0] > payload.start) {
+        // The first gap is between the start and the first timestamp
+        returnArr.push([payload.start, timestamps[0]])
+      }
+      // Iterate over the remaining timestamps and search for gaps
+      for (let index = 1; index < timestamps.length; index += 2) {
+        if (timestamps[index + 1]) {
+          returnArr.push([timestamps[index], timestamps[index + 1]])
+        } else if (timestamps[index] < payload.end) {
+          returnArr.push([timestamps[index], payload.end])
+        }
+      }
+      // Return an array of interval pairs
+      return returnArr
+    }
+  },
+
+  // Retrieves data from the cache if it exists
+  // If the data does not exist, it:
+  //   1. Requests for the data from the API
+  //   2. Caches the datums
+  //   3. Retrieves the data from the cache
+  // Parameters:
+  //  meterId: integer that uniquely identifies the meter
+  //  start: integer representing linux epoch time of the start of the required interval
+  //  end: integer representing linux epoch time of the end of the required interval
+  //  uom: the unit of measure/metering point to request data for
+  //  classInt: An integer that corresponds to the type of meter we are reading from
+  async getData (store, payload) {
+    // Does the cache contain the data?
+    const missingIntervals = await this.dispatch('dataStore/findMissingIntervals', {
+      meterId: payload.meterId,
+      start: payload.start,
+      end: payload.end,
+      uom: payload.uom
+    })
+    if (missingIntervals.length > 0) {
+      // The cache does not contain all of the data
+      // Add it to the cache
+      let promises = []
+      missingIntervals.forEach(async (interval) => { // For each interval, request for the data
+        if (interval.length === 0) interval.push(Math.round(new Date().getTime() / 1000))
```

Comment:
Should this be rounded to nearest 15 min interval?

Reply:
I don't believe it needs to be rounded. This is passed through to the API on line 126, and eventually it will make its way into a SQL query for data where the timestamp is `start <= timestamp <= current time`. If we round it down, technically we should get the same result. If we round it up, we should get the same result (assuming we don't have a meter read up to 15 minutes in the future).

I might need to add a check for small gaps of <15 minutes in _dataStore/findMissingIntervals_ though...

Reply:
I think it will need more exact intervals when doing any of the calculations, it keys off that value being 15 minutes apart. This is probably fine, but maybe make a custom data type just to avoid that dumb issue anyways? Or helper function. I ran into it way too much so I just tried to make calling getTime always go to the floored 15 min interval.

Reply:
@broha22 The intervals contain smaller datum objects that contain their own datetime & meterpoint pairs. All of the actual data points have timestamps rounded to the nearest 15 minute interval for calculations.

These intervals are just used by the caching system to quickly determine what data is "here" and what data is "missing".